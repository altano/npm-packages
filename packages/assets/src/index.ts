import { createRequire } from "node:module";
import { readFile } from "node:fs/promises";

const resolve = createRequire(import.meta.url).resolve;

/**
 * @returns the file path for the Inter woff font file at the given weight
 */
export function getInterPath(fontWeight: number): string {
  return resolve(
    `@fontsource/inter/files/inter-latin-${fontWeight}-normal.woff`,
  );
}

/**
 * @returns the node Buffer for the Inter woff font file at the given weight
 */
export async function getInterBuffer(fontWeight: number): Promise<Buffer> {
  return readFile(getInterPath(fontWeight));
}
