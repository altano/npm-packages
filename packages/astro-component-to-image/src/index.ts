import satori from "satori";
import { defineMiddleware } from "astro/middleware";
import sharp, { type Sharp, type FormatEnum as SharpFormatEnum } from "sharp";
import { html as htmlToVNode } from "satori-html";
import { contentType } from "mime-types";
import he from "he";

import type { MiddlewareResponseHandler, APIContext } from "astro";
import type { SatoriOptions as SatoriOptionsOrig } from "satori";

export type SatoriOptions = SatoriOptionsOrig;
export type SharpOptions = NonNullable<Parameters<Sharp["toFormat"]>[1]>;
export type SharpFormats = keyof SharpFormatEnum;

// TODO Needs lots of tests

export type ImageOptions<Format extends SharpFormats> = {
  format: Format;
  options?: SharpOptions;
};

type Filename<Format extends SharpFormats> = `${string}.${Format}`;

export type Options<Format extends SharpFormats> = {
  /**
   * Any output format that the sharp library accepts, as a string.
   * e.g. "avif", "jpg", "png", "webp", "gif", etc.
   *
   * API documentation: https://sharp.pixelplumbing.com/api-output
   */
  format: Format;
  /**
   * Options that the vercel/satori library accepts. At the very least, you must
   * specify dimensions and one font to use.
   *
   * API documentation:
   * https://github.com/vercel/satori/blob/0a258931fe2291bdd461103780ac01e3c700b845/src/satori.ts#L18-L40
   */
  getSatoriOptions(
    context: APIContext,
    response: Response,
    format: Format,
  ): Promise<SatoriOptions>;
  /**
   * Any output options that the sharp library accepts, e.g.:
   *
   *   { quality: 50 }
   *
   * API documentation: https://sharp.pixelplumbing.com/api-output#toformat
   */
  getSharpOptions?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<SharpOptions>;
  /**
   * This function must return true for the given request or the route will not
   * be converted to an image. By default, only components/endpoints that return
   * a text/html Response and have a route ending in "<name>.<format>", e.g.
   * image.png, will be converted to an image. Use this method to override that
   * behavior and provide your own heuristic.
   */
  shouldReplace?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<boolean>;
  /**
   * The default filename for a component such as `image.png.astro` is
   * `image.png`. You may use this function to override that default. Only
   * relevant in SSG: in SSR there is no filename and Astro just responds with
   * the file image contents.
   */
  getFilename?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<Filename<Format>>;
};

function getContentType<Format extends SharpFormats>(format: Format): string {
  const type = contentType(format);
  return type === false ? `image/${format}` : type;
}

async function defaultGetFilename<Format extends SharpFormats>(
  format: Format,
  context: APIContext,
): Promise<string> {
  const requestUrl = context.url.toString();
  new RegExp(String.raw`\.${format}\/?$`).test(requestUrl);

  const basename = context.url.pathname.split("/").filter(Boolean).at(-1);

  if (basename == null) {
    throw new Error(
      `Could not find basename of URL: ${context.url.toString()}`,
    );
  }

  return basename;
}

async function defaultGetSharpOptions(): Promise<SharpOptions> {
  return { quality: 100 };
}

async function defaultShouldReplace<Format extends SharpFormats>(
  context: APIContext,
  response: Response,
  format: Format,
): Promise<boolean> {
  const contentType = response.headers?.get("content-type");

  // In SSG there will be a trailing slash in the request URL. In SSR/dev a
  // user might type it in without the slash. Handle both.
  const requestUrl = context.url.toString();
  const isImageOfType = new RegExp(String.raw`\.${format}\/?$`).test(
    requestUrl,
  );

  return (
    isImageOfType &&
    // Don't rewrite other things like Endpoints. Only interested in components.
    response instanceof Response &&
    // Don't rewrite anything that isn't html.
    contentType === "text/html"
  );
}

export function createHtmlToImageMiddleware<Format extends SharpFormats>({
  format,
  getSharpOptions,
  shouldReplace,
  getFilename,
  getSatoriOptions,
}: Options<Format>): MiddlewareResponseHandler {
  return defineMiddleware(async (context, next) => {
    const response = await next();
    const replace =
      shouldReplace == null
        ? await defaultShouldReplace(context, response, format)
        : await shouldReplace(context, response, format);

    if (!replace) {
      // This request isn't relevant to us. Don't modify the response.

      // TODO Remove when issue is fixed in astro
      // https://github.com/withastro/astro/issues/8045
      if (isEndpointOutput(response)) {
        return new Response(response.body, {
          status: 200,
          headers: response.headers,
        });
      }

      return response;
    }

    const [responseText, satoriOptions, sharpOptions, filename] =
      await Promise.all([
        response.text(),
        getSatoriOptions(context, response, format),
        getSharpOptions == null
          ? defaultGetSharpOptions()
          : getSharpOptions(context, response, format),
        getFilename == null
          ? defaultGetFilename(format, context)
          : getFilename(context, response, format),
      ]);

    // html text => vnode
    const responseTextWithDecodedHtmlEntities = he.decode(responseText);
    const vnode = htmlToVNode(responseTextWithDecodedHtmlEntities);

    // vnode => svg
    const svg = await satori(vnode as React.ReactNode, satoriOptions);

    // svg => image
    const imageBuffer = await sharp(Buffer.from(svg))
      .toFormat(format, sharpOptions)
      .toBuffer();

    return new Response(imageBuffer, {
      headers: {
        "Content-Type": getContentType(format),
        "Content-Disposition": `inline; filename="${filename}"`,
      },
    });
  });
}

function isEndpointOutput(endpointResult: unknown): boolean {
  return (
    endpointResult != null &&
    !(endpointResult instanceof Response) &&
    typeof endpointResult === "object" &&
    "body" in endpointResult &&
    typeof endpointResult.body === "string"
  );
}
