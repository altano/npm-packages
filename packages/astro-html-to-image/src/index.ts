import satori from "satori";
import { defineMiddleware } from "astro/middleware";
import { Resvg } from "@resvg/resvg-wasm";
import { html as htmlToVNode } from "satori-html";
import { contentType } from "mime-types";
import he from "he";

import type { MiddlewareResponseHandler, APIContext } from "astro";
import type { SatoriOptions } from "satori";
import type { ResvgRenderOptions } from "@resvg/resvg-wasm";

export type SvgOptions = SatoriOptions;
export type ImageRenderOptions = ResvgRenderOptions;
export type ImageFormat = "png";

// TODO Needs lots of tests

export type ImageOptions<Format extends ImageFormat> = {
  format: Format;
  options?: ImageRenderOptions;
};

type Filename<Format extends ImageFormat> = `${string}.${Format}`;

export type Runtime = "nodejs";

export async function initialize({
  runtime,
}: {
  runtime: Runtime;
}): Promise<void> {
  switch (runtime) {
    case "nodejs":
      {
        const initialize = await import("./initializeResvg-Nodejs");
        await initialize.default();
      }
      break;
    default:
      runtime satisfies never;
  }
}

export type Options<Format extends ImageFormat> = {
  /**
   * Your JavaScript runtime, e.g. "nodejs"
   */
  runtime: Runtime;
  /**
   * Any output format that @resvg/resvg-wasm library accepts, as a string.
   * Currently only "png"
   */
  format: Format;
  /**
   * Options that the vercel/satori library accepts. At the very least, you must
   * specify dimensions and one font to use.
   *
   * API documentation:
   * https://github.com/vercel/satori/blob/0a258931fe2291bdd461103780ac01e3c700b845/src/satori.ts#L18-L40
   */
  getSvgOptions(
    context: APIContext,
    response: Response,
    format: Format,
  ): Promise<SvgOptions>;
  /**
   * Any output options that the @resvg/resvg-js library accepts, e.g.:
   *
   *   { background: "rgba(255,255,255, 0.8)" }
   *
   * API documentation: https://github.com/yisibl/resvg-js/blob/main/index.d.ts#L3
   */
  getImageOptions?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<ImageRenderOptions>;
  /**
   * This function must return true for the given request or the route will not
   * be converted to an image. By default, only components/endpoints that return
   * a text/html Response and have a route ending in "<name>.<format>", e.g.
   * image.png, will be converted to an image. Use this method to override that
   * behavior and provide your own heuristic.
   */
  shouldReplace?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<boolean>;
  /**
   * The default filename for a component such as `image.png.astro` is
   * `image.png`. You may use this function to override that default. Only
   * relevant in SSG: in SSR there is no filename and Astro just responds with
   * the file image contents.
   */
  getFilename?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<Filename<Format>>;
};

function getContentType<Format extends ImageFormat>(format: Format): string {
  const type = contentType(format);
  return type === false ? `image/${format}` : type;
}

export async function defaultGetFilename<Format extends ImageFormat>(
  format: Format,
  context: APIContext,
): Promise<string> {
  const requestUrl = context.url.toString();
  new RegExp(String.raw`\.${format}\/?$`).test(requestUrl);

  const basename = context.url.pathname.split("/").filter(Boolean).at(-1);

  if (basename == null) {
    throw new Error(
      `Could not find basename of URL: ${context.url.toString()}`,
    );
  }

  return basename;
}

export async function defaultGetImageOptions(): Promise<ImageRenderOptions> {
  return {
    textRendering: 1, // optimizeLegibility
    imageRendering: 0, // optimizeQuality
  };
}

export async function defaultShouldReplace<Format extends ImageFormat>(
  context: APIContext,
  response: Response,
  format: Format,
): Promise<boolean> {
  const contentType = response.headers?.get("content-type");

  // In SSG there will be a trailing slash in the request URL. In SSR/dev a
  // user might type it in without the slash. Handle both.
  const requestUrl = context.url.toString();
  const isImageOfType = new RegExp(String.raw`\.${format}\/?$`).test(
    requestUrl,
  );

  return (
    isImageOfType &&
    // Don't rewrite other things like Endpoints. Only interested in components.
    response instanceof Response &&
    // Don't rewrite anything that isn't html.
    contentType === "text/html"
  );
}

let isWasmInitialized = false;

export function createHtmlToImageMiddleware<Format extends ImageFormat>({
  runtime,
  format,
  getImageOptions,
  shouldReplace,
  getFilename,
  getSvgOptions,
}: Options<Format>): MiddlewareResponseHandler {
  return defineMiddleware(async (context, next) => {
    if (!isWasmInitialized) {
      isWasmInitialized = true;
      try {
        await initialize({ runtime });
      } catch (err) {
        // Restarting the dev server causes us to unnecessarily initialize the
        // wasm module twice. Let's just ignore the error (since we can't
        // prevent it).
        const shouldIgnore =
          err instanceof Error &&
          err.message?.startsWith("Already initialized");
        if (!shouldIgnore) {
          // re-throw all other errors
          throw err;
        }
      }
    }

    const response = await next();
    const replace =
      shouldReplace == null
        ? await defaultShouldReplace(context, response, format)
        : await shouldReplace(context, response, format);

    if (!replace) {
      // This request isn't relevant to us. Don't modify the response.

      // TODO Remove when issue is fixed in astro
      // https://github.com/withastro/astro/issues/8045
      if (isEndpointOutput(response)) {
        return new Response(response.body, {
          status: 200,
          headers: response.headers,
        });
      }

      return response;
    }

    const [responseText, svgOptions, imageOptions, filename] =
      await Promise.all([
        response.text(),
        getSvgOptions(context, response, format),
        getImageOptions == null
          ? defaultGetImageOptions()
          : getImageOptions(context, response, format),
        getFilename == null
          ? defaultGetFilename(format, context)
          : getFilename(context, response, format),
      ]);

    // html text => vnode
    const responseTextWithDecodedHtmlEntities = he.decode(responseText);
    const vnode = htmlToVNode(responseTextWithDecodedHtmlEntities);

    // vnode => svg
    const svg = await satori(vnode as React.ReactNode, svgOptions);

    // svg => image
    const imageResvgArr = new Resvg(svg, imageOptions).render().asPng();
    const imageBuffer = Buffer.from(imageResvgArr);

    return new Response(imageBuffer, {
      headers: {
        "Content-Type": getContentType(format),
        "Content-Disposition": `inline; filename="${filename}"`,
      },
    });
  });
}

function isEndpointOutput(endpointResult: unknown): boolean {
  return (
    endpointResult != null &&
    !(endpointResult instanceof Response) &&
    typeof endpointResult === "object" &&
    "body" in endpointResult &&
    typeof endpointResult.body === "string"
  );
}
