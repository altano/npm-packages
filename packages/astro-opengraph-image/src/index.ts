import { createImageMiddleware } from "./createImageMiddleware";

export const createOpenGraphImageMiddleware = createImageMiddleware;
