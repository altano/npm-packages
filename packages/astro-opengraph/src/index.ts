export { default } from "./integration.js";
export type {
  OpengraphImageConfig,
  ImageOptionsWithFontPaths,
  ComponentMetaTagFallbacks,
} from "./types.js";
