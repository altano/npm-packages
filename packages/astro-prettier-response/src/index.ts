export { default } from "./integration.js";
export type { AstroPrettierResponseConfig } from "./config.js";
