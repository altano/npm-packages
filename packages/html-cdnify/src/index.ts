export * from "./cdnify.js";
export * from "./CDNTransformer.js";
