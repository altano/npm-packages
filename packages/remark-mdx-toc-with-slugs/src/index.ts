import { generate } from "astring";
import { valueToEstree } from "estree-util-value-to-estree";
import { EXIT, visit as estreeVisit } from "estree-util-visit";
import GithubSlugger from "github-slugger";
import { remarkMdxToc } from "remark-mdx-toc";
import { visit } from "unist-util-visit";

import type { RemarkMdxTocOptions, TocEntry } from "remark-mdx-toc";
import type { Plugin } from "unified";
import type { MdxjsEsm } from "mdast-util-mdxjs-esm";
import type { Node } from "unist-util-visit/lib";
import type { ArrayExpression } from "estree";

type TableOfContentsEntry = TocEntry & {
  slug: string;
  children: TableOfContentsEntry[];
};

function sluggifyTocEntries(
  entries: TocEntry[],
  headingSlugger = new GithubSlugger(),
): TableOfContentsEntry[] {
  return entries.map((entry) => sluggifyTocEntry(entry, headingSlugger));
}

function sluggifyTocEntry(
  entry: TocEntry,
  headingSlugger: GithubSlugger,
): TableOfContentsEntry {
  return {
    ...entry,
    slug: headingSlugger.slug(entry.value),
    children: sluggifyTocEntries(entry.children, headingSlugger),
  };
}

export type RemarkMdxTocWithSlugsOptions = RemarkMdxTocOptions;

function assertMdxjsEsmNode(node: Node): asserts node is MdxjsEsm {
  if (node.type !== "mdxjsEsm") {
    throw new Error(`Node was unexpected type: ${node.type}`);
  }
}

/**
 * Remark plugin that, given a tree that has had remark-mdx-toc run over it
 * already, the toc data will be enhanced to have auto-generated slugs for every
 * heading. This uses github-slugger
 */
const remarkMdxTocWithSlugs: Plugin<[RemarkMdxTocWithSlugsOptions?]> = (
  options = {},
) => {
  const name = options.name ?? "toc"; // mirrors remark-mdx-toc

  return function mdxTocWithSlugsTransformer(ast): void {
    const mdast = ast;

    // visit all esm nodes in the mdx
    visit(mdast, ["mdxjsEsm"], (mdxNode) => {
      assertMdxjsEsmNode(mdxNode);
      // visit all nodes in the estree
      const estree = mdxNode.data?.estree;
      if (estree) {
        estreeVisit(estree, (esNode) => {
          if (
            esNode.type === "VariableDeclarator" &&
            esNode.id.type === "Identifier" &&
            esNode.id.name === name
          ) {
            // we found the "toc" variable declarator
            if (esNode.init) {
              // It's easier to be wasteful and do our transformations in JS,
              // instead of directly on the tree, and then convert it back to a
              // tree.
              const tocExportString = generate(esNode.init);
              const toc = JSON.parse(tocExportString);

              // Add slugs to the toc
              const tocWithSlugs = sluggifyTocEntries(toc);

              // Overwrite the existing toc export with ours
              const newTree = valueToEstree(tocWithSlugs);

              if (newTree.type === "ArrayExpression") {
                esNode.init = newTree as ArrayExpression; // different versions of @types/estree are screwing with TypeScript
              }
            }

            // We out
            return EXIT;
          }

          return;
        });
      }
    });
  };
};

// This plugin is a preset that combines remark-mdx-toc and our enhancements
// plugin. This makes it so that the user doesn't have to separately add the
// remark-mdx-toc plugin themselves.
const preset = {
  plugins: [remarkMdxToc, remarkMdxTocWithSlugs],
};

export default preset;
