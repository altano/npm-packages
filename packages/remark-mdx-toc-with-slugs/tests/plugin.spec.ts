import plugin from "../src/index";
import { describe } from "vitest";
import { testByFixtures } from "@altano/remark-plugin-test-util/src/testByFixture";

describe("remark-mdx-toc-with-slugs", async () => {
  await testByFixtures(plugin);
});
