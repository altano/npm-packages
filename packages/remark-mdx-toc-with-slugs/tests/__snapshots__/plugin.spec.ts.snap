// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remark-mdx-toc-with-slugs > [test #0] basic 1`] = `
/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const toc = [
  {
    depth: 2,
    value: "Resource Acquisition Is Initialization (RAII)",
    attributes: {},
    children: [
      {
        depth: 3,
        value: "Is There a Problem?",
        attributes: {},
        children: [],
        slug: "is-there-a-problem",
      },
      {
        depth: 3,
        value: "C++ Classes",
        attributes: {},
        children: [],
        slug: "c-classes",
      },
      {
        depth: 3,
        value: "The Solution",
        attributes: {},
        children: [],
        slug: "the-solution",
      },
    ],
    slug: "resource-acquisition-is-initialization-raii",
  },
];
function _createMdxContent(props) {
  const _components = {
    h2: "h2",
    h3: "h3",
    ...props.components,
  };
  return (
    <>
      <_components.h2>
        {"Resource Acquisition Is Initialization (RAII)"}
      </_components.h2>
      {"\\n"}
      <_components.h3>{"Is There a Problem?"}</_components.h3>
      {"\\n"}
      <_components.h3>{"C++ Classes"}</_components.h3>
      {"\\n"}
      <_components.h3>{"The Solution"}</_components.h3>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}

`;

exports[`remark-mdx-toc-with-slugs > [test #1] custom-name 1`] = `
/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const tableOfContents = [
  {
    depth: 2,
    value: "Resource Acquisition Is Initialization (RAII)",
    attributes: {},
    children: [
      {
        depth: 3,
        value: "Is There a Problem?",
        attributes: {},
        children: [],
        slug: "is-there-a-problem",
      },
      {
        depth: 3,
        value: "C++ Classes",
        attributes: {},
        children: [],
        slug: "c-classes",
      },
      {
        depth: 3,
        value: "The Solution",
        attributes: {},
        children: [],
        slug: "the-solution",
      },
    ],
    slug: "resource-acquisition-is-initialization-raii",
  },
];
function _createMdxContent(props) {
  const _components = {
    h2: "h2",
    h3: "h3",
    ...props.components,
  };
  return (
    <>
      <_components.h2>
        {"Resource Acquisition Is Initialization (RAII)"}
      </_components.h2>
      {"\\n"}
      <_components.h3>{"Is There a Problem?"}</_components.h3>
      {"\\n"}
      <_components.h3>{"C++ Classes"}</_components.h3>
      {"\\n"}
      <_components.h3>{"The Solution"}</_components.h3>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}

`;
